numba/typing/templates.py

```diff
@@ -113,6 +113,13 @@ def as_function(self):
         sig = signature(self.return_type, *((self.recvr,) + self.args))
         return sig
 
+    def __call__(self, *args):
```

Comment:
Why is this necessary? Signature can't really numba type.

Reply:
This is a convenience method for constructing signatures that involve function types as return values. (When trying it now, I see it is broken but) as an example, consider
```
from numpy import *
int32(int32)   # this results a signature `(int64,) -> int64`
```
With the first-class function types, we also need to construct a signature where the return type is a function type. For instance, it would be convenient to construct the corresponding signature as follows 
```
rtype = int32(int32)
sig = rtype(int32)           # here the __call__ methods is effective
```
that would be a signature of a function taking int32 value as input and returning a function of type `int32(int32)`.

It is in some sense unfortunate that `int32(int32)` returns a Signature object. With the first-class function support, it would be more natural if it would return numba `FunctionType` instance.

Reply:
> it would be more natural if it would return numba FunctionType instance.

Agree, we need to look at that as part of our Q1 housekeeping.

Reply:
Removed `__call__` method for now.
TODO for Q1 or later: `Type.__call__` returns `TypeFunction` instance. The question of how this can be transformed to `Signature` instance, needs an answer. E.g. is `Signature` still required as `TypeFunction` carries the same info?